torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -449,6 +476,101 @@ void ProcessGroupAgent::listenLoop() {
   }
 }
 
+void ProcessGroupAgent::pollTimedOutRPCs() {
+  // Sleep until a future is available, or there are no futures and we are
+  // shutting down.
+  {
+    std::unique_lock<std::mutex> lock(futureTimeoutMutex_);
+    futureTimeoutCV_.wait(lock);
+  }
+
+  while (!shutdown_.load()) {
+    std::chrono::milliseconds sleepTime;
+    {
+      std::lock_guard<std::mutex> lock{futureMutex_};
+      // Estimate amount of time the first future will time out in, and sleep
+      // for that long.
+      if (futureTimeouts_.empty()) {
+        continue;
+      }
+      auto runningTime =
+          std::chrono::duration_cast<std::chrono::milliseconds>(
+              std::chrono::steady_clock::now().time_since_epoch()) -
+          futureTimeouts_.begin()->first;
+      // TODO: will sleep for 0 ms when rpcTimeout_ is 0 (meaning infinite),
+      // which will result in busy polling for the lock If rpcTimeout_ is 0,
+      // should we have a default sleep interval here?
+      sleepTime =
+          std::max(rpcTimeout_ - runningTime, std::chrono::milliseconds(0));
+    }
+
+    // Sleep for the ETA for the next future to time out, or until we are told
+    // to shut down. This is to avoid the case where we sleep for a long time
+    // due to a future with a high timeout.
+    {
+      std::unique_lock<std::mutex> lock(futureTimeoutMutex_);
+      if (futureTimeoutCV_.wait_for(
+              lock, sleepTime, [this] { return shutdown_.load(); })) {
```

Comment:
@pietern, let me know if this wasn't exactly what you were suggesting. Also, I think we'd need the sleep to be interruptible, since we wouldn't want to sleep if process group agent is shutting down.

Reply:
You might not want the predicate looping here, but rather just wait_for (or wait_until)
The reason is because if there's a newly added entry to the table with a sooner expiration time, you want to wake up and recalculate

Reply:
I expected to collapse the top block and this one into a single one. The timeouts are stored in the future timeout map, so why do we need access to the futures themselves to figure out how long we should sleep?

Reply:
@jjlilley Are we concerned that a lot of futures will be created (with say, the same timeout) and thus this thread would be woken up too frequently? This was the reasoning why I had that predicate.

Reply:
@rohan-varma 
The thread gets woken up the same number of times with/without the predicate...

The predicate ends up just being a while loop syntactic sugar.

In the actual generated code, it is [C pseudo-code] something like 

   if (!pred()) {
     auto absolute_timeout = now() + timeout;
     do {
        if (pthread_cond_timedwait(&cv, &lock->mutex, absolute_timeout) == -1 && errno == ETIMEOUT)  break;
      } while (!pred());
  }

Hence, adding a predicate doesn't really "save" anything, and you might as well check if you've hit an expiration after the cv wakes up the thread. And in this particular case, having the pred was preventing it from waking up from an infinite timeout, if you wanted one for 0 futures.

If you have a grouping of futures with effectively the exact same absolute timeout, your logic should effectively be able to expire the all at once anyway, so it's not like there's a single wakeup per future. You really only need to notify the condvar if you
  - go from 0 futures to 1
  - insert an entry with a timeout lower than your current earliest timeout
  - shut down

Reply:
@pietern The two different blocks are now collapsed in the latest update.

Reply:
@jjlilley right now we don't support per-RPC timeouts, so the 2nd condition can't happen. I changed it so that we only notify in the 1st and 3rd conditions you mentioned.